core/src/main/scala/org/apache/spark/sql/delta/commands/RestoreTableCommand.scala

```diff
@@ -16,113 +16,128 @@
           .history
-          .getActiveCommitAtTime(timestamp.get, canReturnLastCommit = true)
+          .getActiveCommitAtTime(parseStringToTs(timestamp), canReturnLastCommit = true)
           .version
-      )
+      }
 
-      require(versionToRestore < latestSnapshot.version,
-        s"Version to restore ($versionToRestore) should be less then " +
-        s"last available version (${latestSnapshot.version})")
+      val latestVersion = deltaLog.update().version
 
-      val snapshotToRestore = deltaLog.getSnapshotAt(versionToRestore)
-      val latestSnapshotFiles = latestSnapshot.allFiles
-      val snapshotToRestoreFiles = snapshotToRestore.allFiles
+      require(versionToRestore < latestVersion, s"Version to restore ($versionToRestore)" +
+        s"should be less then last available version ($latestVersion)")
 
-      import spark.implicits._
-      import collection.JavaConverters._
-
-      val filesToAdd = snapshotToRestoreFiles
-        .join(
-          latestSnapshotFiles,
-          snapshotToRestoreFiles("path") ===  latestSnapshotFiles("path"),
-          "left_anti")
-        .as[AddFile]
-        .map(_.copy(dataChange = true))
-        .repartition(parallelism)
-        .cache() // To avoid Dataset recompute for each partition of toLocalIterator()
-
-      checkSnapshotFilesAvailability(deltaLog, filesToAdd, versionToRestore)
-
-      val filesToRemove = latestSnapshotFiles
-        .join(
-          snapshotToRestoreFiles,
-          latestSnapshotFiles("path") ===  snapshotToRestoreFiles("path"),
-          "left_anti")
-        .as[AddFile]
-        .map(_.removeWithTimestamp())
-        .repartition(parallelism)
-        .cache() // To avoid Dataset recompute for each partition of toLocalIterator()
-
-      // Commit files, metrics, protocol and metadata to delta log
       deltaLog.withNewTransaction { txn =>
-
-        val metrics = computeMetrics(filesToAdd, filesToRemove, snapshotToRestore)
-        val addActions = filesToAdd.toLocalIterator().asScala
-        val removeActions = filesToRemove.toLocalIterator().asScala
-
-        txn.updateMetadata(snapshotToRestore.metadata)
-
-        commitLarge(
-          spark,
-          txn,
-          addActions ++ removeActions,
-          DeltaOperations.Restore(version, timestamp.map(_.getTime)),
-          Map.empty,
-          metrics)
+        val latestSnapshot = txn.snapshot
+        val snapshotToRestore = deltaLog.getSnapshotAt(versionToRestore)
+        val latestSnapshotFiles = latestSnapshot.allFiles
+        val snapshotToRestoreFiles = snapshotToRestore.allFiles
+
+        import spark.implicits._
+
+        val filesToAdd = snapshotToRestoreFiles
+          .join(
+            latestSnapshotFiles,
+            snapshotToRestoreFiles("path") === latestSnapshotFiles("path"),
+            "left_anti")
+          .as[AddFile]
+          .map(_.copy(dataChange = true))
+          .cache() // To avoid Dataset recompute for each partition of toLocalIterator()
```

Comment:
why would localIterator cause recomputation. local iterator would compute partition one by one, but that should not cause any reprocessing of previously processed partitions.

Reply:
I agree that comment was really confusing. I've updated comment to make it more correct and clear